File: python/game.py
import pygame,math,random,time
storageFile = "stackBox.csv"
pygame.init()
RunGame = True
device = pygame.display.Info()
screen = pygame.display.set_mode((device.current_w,device.current_h))
clock = pygame.time.Clock()
fontFamily = '../Grandstander/static/Grandstander-Medium.ttf'
running  = True
dt = 0
first_block = True
#gradiant code
def gradientRect( window, left_colour, right_colour, target_rect ):
    colour_rect = pygame.Surface( ( 2, 2 ) )                                   # tiny! 2x2 bitmap
    pygame.draw.line( colour_rect, left_colour,  ( 0,0 ), ( 1,0 ) )            # left colour line
    pygame.draw.line( colour_rect, right_colour, ( 0,1 ), ( 1,1 ) )            # right colour line
    colour_rect = pygame.transform.smoothscale( colour_rect, ( target_rect.width, target_rect.height ) )  # stretch!
    window.blit( colour_rect, target_rect )  
#blocks
class Game:
    width = 500
    height = device.current_h
    x= device.current_w*0.5-width*0.5
    y = 0
class coords:
    x =0
    y =0
    width =10
    height = 10
    
class Coners:
    def __init__(self):
        self.top =coords()
        self.bottom =coords()
        self.left =coords()
        self.right =coords()
class Float_Text:
    def __init__(self,text):
        self.y = device.current_h/2
        self.opacity = 1
        self.text = text
    
    def draw(self):
        color = pygame.Color(1)
        color.hsla = 2,100,50,100
        font = pygame.font.Font(fontFamily,80)
        text = font.render(self.text,True,(0,0,0))
        textRect = text.get_rect()
        textRect.centerx =device.current_w//2
        textRect.top = self.y
        text = text.convert_alpha()
        text.fill((255, 255, 255, int(self.opacity*255)), special_flags=pygame.BLEND_RGBA_MULT)
        screen.blit(text,textRect)
    
    def update(self):
        self.y-=1
        self.opacity-=0.005
float_texts = []
def create_text (text):
    float_texts.append(Float_Text(text))
def handle_text ():
    for txt in float_texts:
        txt.draw()
        txt.update()
        if(txt.opacity <= 0):
            float_texts.remove(txt)
class times:
    start = 0
    end = 0
class Blocks:
    start_msg = True;msg = 'Click to start\nor\nPress "Space Bar"\n';start= True;x = device.current_w/2;y = device.current_h-250;dirX = -1 if math.floor(random.random()*2) else 1;dirY =  -1 if math.floor(random.random()*2) else 1;hsl = 60;width = 200;height = 200;speed = 0.2;score = 0;timer = "00:00";highscore = 0;time = times()
    def __init__(self):
        self.hsl = Blocks.hsl
        self.x = Blocks.x
        self.y = Blocks.y
        self.dirX = Blocks.dirX
        self.dirY = Blocks.dirY
        self.run = True
        self.width = Blocks.width
        self.height = Blocks.height
        self.length = 50
        self.speed = Blocks.speed
        self.coner= Coners()
        global first_block
        if(not first_block):
            self.update(-(Game.width)*Game.width/Game.height)
        else:
            first_block = False
            Blocks.y-=self.length*2
        
    def draw(self):
        x = self.x; y = self.y;angle = 30
        color = pygame.Color(0)
        color.hsla=(self.hsl,100,50,100)
        polygon = []
        self.coner.top.x = x-5
        self.coner.top.y = y-5
        polygon.append([x,y])
        (x,y) = self.rotateLine(x,y,angle,self.width)
        polygon.append([x,y])
        angle+=120
        (x,y) = self.rotateLine(x,y,angle,self.height)
        polygon.append([x,y])
        angle+=60
        (x,y) = self.rotateLine(x,y,angle,self.width)
        polygon.append([x,y])
        angle+=120
        (x,y) = self.rotateLine(x,y,angle,self.height)
        polygon.append([x,y])
        pygame.draw.polygon(screen,color ,polygon)
        polygon = []
        color.hsla = (self.hsl,100,40,100)
        angle -=180
        (x,y) = self.rotateLine(x,y,angle,self.height)
        polygon.append([x,y])
        angle-=120
        (x,y) = self.rotateLine(x,y,angle,self.width)
        polygon.append([x,y])
        angle+=60
        (x,y) = self.rotateLine(x,y,angle,self.length)
        polygon.append([x,y])
        angle+=120
        (x,y) = self.rotateLine(x,y,angle,self.width)
        self.coner.left.x = x-5
        self.coner.left.y = y-5
        polygon.append([x,y])
        angle+=60
        (x,y) = self.rotateLine(x,y,angle,self.length)
        polygon.append([x,y])
        pygame.draw.polygon(screen,color ,polygon)
        color.hsla = (self.hsl,100,70,100)
        polygon = []
        angle -=240
        (x,y) = self.rotateLine(x,y,angle,self.width)
        polygon.append([x,y])
        angle-=60
        (x,y) = self.rotateLine(x,y,angle,self.height)
        polygon.append([x,y])
        angle+=120
        (x,y) = self.rotateLine(x,y,angle,self.length)
        self.coner.right.x = x-5
        self.coner.right.y = y-5
        polygon.append([x,y])
        angle+=60
        (x,y) = self.rotateLine(x,y,angle,self.height)
        self.coner.bottom.x = x-5
        self.coner.bottom.y = y-5
        polygon.append([x,y])
        angle+=120
        (x,y) = self.rotateLine(x,y,angle,self.length)
        polygon.append([x,y])
        pygame.draw.polygon(screen,color ,polygon)
        
    def rotateLine( self,x,y,angle,len):
        rotate= angle *math.pi/180
        x+=len*math.cos(rotate)
        y+=len*math.sin(rotate)
        return (x,y)
    
    def update(self,pos = 1):
        if (self.run) :
            if(self.coner.right.x > Game.x+Game.width and pos == 1):
                self.dirX=-1 if (self.dirX==self.dirY) else 1;self.dirY=-1
            if(self.coner.left.x <Game.x and pos==1):
                self.dirX=1 if (self.dirX==self.dirY) else -1;self.dirY=1
            rotate =30 if (self.dirY == 1) else 150
            angle =self.dirX*rotate
            self.x+=math.cos(math.radians(angle))*pos*self.speed
            self.y+=math.sin(math.radians(angle))*pos*self.speed    
    def stop(self):
        self.speed = 0
        if self.run: 
            Blocks.y+=self.length
        self.run = False
        if(Blocks.y <= 200):
            self.y+=self.length
    def changeSize(self,b):
        if(math.floor(self.x) == math.floor(b.x) or math.floor(self.y) ==math.floor(b.y)):
            if(Blocks.score):
                create_text("PERFECT")
        Blocks.score += (5 if (math.floor(self.x) == math.floor(b.x) or math.floor(self.y) == math.floor(b.y)) else 1)
        if(b.coner.top.x < self.coner.top.x and b.coner.right.x < self.coner.right.x and self.dirX!=self.dirY):
            self.height -= abs(math.floor(self.x)-b.x)
            self.x =b.x
            self.y = b.y-self.length
            Blocks.height = self.height
        
        if(b.coner.top.x > self.coner.top.x and b.coner.right.x > self.coner.right.x and self.dirX!=self.dirY):
            self.height -= abs(math.floor(self.x)-b.x)
            self.draw()
            b.draw()
            dx = abs(self.x - self.coner.left.x)
            dx1 = abs(b.x - b.coner.left.x)
            self.x = b.x-(dx1-dx)
            self.draw()
            b.draw()
            dx2 = abs(b.coner.left.y - self.coner.left.y)
            self.y+=dx2-self.length
            Blocks.height = self.height
         
        if(b.coner.top.x > self.coner.top.x and b.coner.right.x > self.coner.right.x and self.dirX==self.dirY):
            self.width -= abs(math.floor(self.x)-b.x)
            self.x =b.x
            self.y = b.y-self.length
            Blocks.width = self.width
        
        if(b.coner.top.x < self.coner.top.x and b.coner.right.x < self.coner.right.x and self.dirX==self.dirY):
            self.width -= abs(math.floor(self.x)-b.x)
            self.draw()
            b.draw()
            dx = abs(self.x - self.coner.right.x)
            dx1 = abs(b.x - b.coner.right.x)
            self.x = b.x+abs(dx1-dx)
            Blocks.width = self.width
            self.draw()
            b.draw()
            dx2 = abs(b.coner.right.y - self.coner.right.y)
            self.y+=dx2-self.length  
        Blocks.x = self.x
        Blocks.y = self.y-self.length
        global RunGame
        if(self.height <= 0 or self.width <= 0 and RunGame):
            Blocks.speed = 0
            Blocks.msg= "Click to Restart\n"
            create_text("Game Over")
            RunGame = False
        
with open(storageFile,"r") as file:
    reader = file.read()
    Blocks.highscore = int(reader)
    
            
blocks = []
def create_blocks():
    Blocks.dirX = -1 if math.floor(random.random()*2) else 1
    Blocks.dirY = -1 if math.floor(random.random()*2) else 1
    blocks.append(Blocks())

def handle_blocks():
    for block in blocks:
        block.draw()
        block.update()
        if(block.height <= 0 or block.width <=0):
            blocks.remove(block)
        
#click
def blocks_stop ():
    for block in blocks :
        block.stop()
    if(blocks[len(blocks)-2] and blocks[len(blocks)-1]):
        blocks[len(blocks)-1].changeSize(blocks[len(blocks)-2]) 
create_blocks()
blocks_stop()
Blocks.score = 0
create_blocks()
def onClick():
    global blocks,first_block,RunGame
    if(not RunGame):
        RunGame=True
        blocks = []
        first_block = False
        Blocks.speed = 0.2
        Blocks.score = 0
        Blocks.start = True
        Blocks.start_msg = True
        Blocks.width = 200
        Blocks.height = 200
        Blocks.x=device.current_w/2
        Blocks.y=Game.height-250
        Blocks.hsl = 60
        create_blocks()
        blocks_stop()
        Blocks.score = 0
        create_blocks()
        return

    Blocks.hsl+=1
    Blocks.speed+=0.0005
    blocks_stop()
    if(Blocks.height > 0 or Blocks.width > 0):
        if(RunGame):
            create_blocks()
    msg = ["Nice","Cool","Good","Wow","Super"]
    if(Blocks.start):
        Blocks.time.start = int(time.time())
        Blocks.start=False
    if(Blocks.score%25 == 0):
        create_text(msg[math.floor(random.random()*len(msg))])
    if(int(Blocks.highscore) < Blocks.score):
        Blocks.show_msg = False
        Blocks.highscore = Blocks.score
        if(Blocks.show_msg):
            create_text("New Score")
        
     
     
#animation

def log(msg):
    word = ""
    h = 0
    for i in msg:
        if i == "\n" or i == msg[len(msg)-1]:
            font = pygame.font.Font(fontFamily,30)
            text = font.render(str(word), True,(51,51,51))
            textRect = text.get_rect()
            textRect.center=device.current_w//2,device.current_h//2
            textRect.top += h
            h+=30
            word = ""
            screen.blit(text, textRect)
            continue
        word+=i
def score_num():
    font = pygame.font.Font(fontFamily,80)
    text = font.render(str(Blocks.score), True,(51,51,51))
    textRect = text.get_rect()
    textRect.centerx =device.current_w//2
    textRect.top = 20
    screen.blit(text, textRect)
def highscore():
    font = pygame.font.Font(fontFamily,15)
    text = font.render("HighScore: "+str(Blocks.highscore), True,(51,51,51))
    textRect = text.get_rect()
    textRect.top = 10
    textRect.left = 10
    screen.blit(text, textRect)
def speed():
    font = pygame.font.Font(fontFamily,15)
    text = font.render("Speed: "+str(math.floor((Blocks.speed/0.2)*10)/10)+"x", True,"black")
    textRect = text.get_rect()
    textRect.top = 10
    textRect.right = 10
    textRect.left = device.current_w-120
    screen.blit(text, textRect)
def scoretext():
    font = pygame.font.Font(fontFamily,15)
    text = font.render("Score", True,(51,51,51))
    textRect = text.get_rect()
    textRect.top = 90
    textRect.centerx =device.current_w//2
    screen.blit(text, textRect)
def timer():
    font = pygame.font.Font(fontFamily,15)
    text = font.render(str(Blocks.timer), True,(51,51,51))
    textRect = text.get_rect()
    textRect.top = 120
    textRect.centerx =device.current_w//2
    screen.blit(text, textRect)
def showFPS(fps):
    font = pygame.font.Font(fontFamily,15)
    text = font.render("FPS: "+str(fps), True,(51,51,51))
    textRect = text.get_rect()
    textRect.top = 50
    textRect.left = 10
    screen.blit(text, textRect)
def setTimer ():
    if(Blocks.time.start and RunGame):
        Blocks.time.end = int(time.time())
        sec = math.floor((Blocks.time.end - Blocks.time.start))
        min = math.floor(sec/60)
        txtsec = str(sec % 60);txtmin = str(min)
        if(sec%60 < 10):
            txtsec= "0"+str(sec%60)
        if(min<10):
            txtmin= "0"+str(min)
        Blocks.timer = txtmin+":"+txtsec
frame = 0
start_time = int(time.time())
frame_txt = 0
while running:
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            running = False
        if event.type == pygame.KEYDOWN: 
            if event.key == pygame.K_SPACE:
                Blocks.msg = ""
                onClick() 
                break
            if event.key == pygame.K_ESCAPE:
                running = False
        if event.type == pygame.MOUSEBUTTONDOWN:
            if event.button == 1:  # Left mouse button
                Blocks.msg = ""
                onClick() 
                break
        elif event.type == pygame.FINGERDOWN:
            Blocks.msg = ""
            onClick() 
            break
    screen.fill("black")
    gradientRect( screen, "skyblue", "lightblue", pygame.Rect( 0,0, device.current_w,device.current_h ) )
    handle_blocks()
    handle_text()
    score_num()
    highscore()
    scoretext()
    if(Blocks.speed):
        speed()
    setTimer()
    timer()
    log(Blocks.msg)
    if int(time.time()) - start_time >= 1.0:
        start_time = int(time.time())
        frame_txt = frame
        frame = 0
    frame+=1
    showFPS(frame_txt)
    pygame.display.flip()
   
pygame.QUIT
with open(storageFile,"w") as file:
    write = file.write(str(Blocks.highscore))

File: javascript/game.js
/**@type {HTMLCanvasElement} */
const score_num= document.getElementById("score_num")
const timer= document.getElementById("timer")
const highscore= document.getElementById("highscore")
const speed= document.getElementById("speed")
const messages= document.getElementById("msg")
const FPS = document.getElementById("fps")
///////////////////
let frame = 0;
let start_time = new Date().getTime();
const c = document.getElementById("game");
const ctx = c.getContext("2d");
c.width = 500;c.height = window.innerHeight;
let first_block = true
const abs =x=>x>=0?x:-x
const Block = {
    x:c.width/2,
    y:c.height-250,
    width:200,
    height:200,speed:0.2,
    score:0,start:false,
    time:{
        start:0,
        end:0
    },
    highscore:localStorage.getItem("highscore")||0,
    show_msg:true
}

let RunGame = true;
class Float_Text{
    constructor(text){
        this.x = c.width/2;
        this.y = c.height/2
        this.opacity = 1
        this.text = text
    }
    draw(){
        ctx.beginPath();
        ctx.fillStyle =`rgba(0,0,0,${this.opacity})`
        ctx.textAlign = "center";
        ctx.font = '80px "Grandstander"'
        ctx.fillText(this.text,this.x,this.y,)
        ctx.fill()
        ctx.closePath();
    }
    update(){
        this.y--
        this.opacity-=0.005;

    }
}
let float_texts = []
const create_text= (text)=>{
    float_texts.push(new Float_Text(text))
}
const handle_text = ()=>{
    for (let i = 0; i < float_texts.length; i++) {
        const txt = float_texts[i];
        txt.draw();
        txt.update()
        if(txt.opacity <= 0){
            float_texts.splice(i,1);
            i--;
        }
    }
}
class Blocks{
    constructor(x,y,color){
        this.hsl = color;
        this.x = Block.x
        this.y = Block.y 
        const portrait = c.width/c.height,lanscape = c.height/c.width;
        this.coner = {
            top:{
                x:0,y:0,s:10
            },
            bottom:{
                x:0,y:0,s:10
            },
            left:{
                x:0,y:0,s:10
            },
            right:{
                x:0,y:0,s:10
            }
        }
        
        this.run = true  
        this.dirX = x,this.dirY =y
        // 
        this.width = Block.width;
        this.height = Block.height;
        this.color = color
        this.length = 50;
      
        this.speed = Math.floor(Block.speed*10)/10;
        if(!first_block){
            this.update(-c.width*portrait)
            first_block = false
            
        }else{
            Block.y-=this.length*2
            first_block = false
        } 
    }
    draw(){
        let x = this.x, y = this.y,getXY,angle = 30

        ///
        ctx.beginPath();
        ctx.fillStyle=`hsl(${this.hsl},100%,50%)`
        this.coner.top.x = x-5;
        this.coner.top.y = y-5
        ctx.moveTo(x,y);
        getXY = this.rotateLine(x,y,angle,this.width);x=getXY[0],y=getXY[1]
        ctx.lineTo(x,y);
        angle+=120
        getXY = this.rotateLine(x,y,angle,this.height);x=getXY[0],y=getXY[1]
        
        ctx.lineTo(x,y);
        angle+=60
        getXY = this.rotateLine(x,y,angle,this.width);x=getXY[0],y=getXY[1]
        ctx.lineTo(x,y);
        angle+=120
        getXY = this.rotateLine(x,y,angle,this.height);x=getXY[0],y=getXY[1]
        ctx.lineTo(x,y);
        ctx.fill()
        ctx.closePath();
        /////
         
        ////////////
        ctx.beginPath();
        
        angle-=180
        getXY = this.rotateLine(x,y,angle,this.height);x=getXY[0],y=getXY[1]
        ctx.fillStyle = `hsl(${this.hsl},100%,40%)`
        ctx.moveTo(x,y);
        angle-=120
        getXY = this.rotateLine(x,y,angle,this.width);x=getXY[0],y=getXY[1]
        ctx.lineTo(x,y);
        angle+=60
        getXY = this.rotateLine(x,y,angle,this.length);x=getXY[0],y=getXY[1]
        ctx.lineTo(x,y);
        angle+=120
        getXY = this.rotateLine(x,y,angle,this.width);x=getXY[0],y=getXY[1]
        this.coner.left.x = x-5;
        this.coner.left.y = y-5
        ctx.lineTo(x,y);
        angle+=60
        getXY = this.rotateLine(x,y,angle,this.length);x=getXY[0],y=getXY[1]
        ctx.lineTo(x,y);
        ctx.fill()
        ctx.closePath();
        /////////////////
        ctx.beginPath();
        ctx.fillStyle =`hsl(${this.hsl},100%,70%)`
        angle-=240
        getXY = this.rotateLine(x,y,angle,this.width);x=getXY[0],y=getXY[1]
        ctx.moveTo(x,y)
        angle-=60
        getXY = this.rotateLine(x,y,angle,this.height);x=getXY[0],y=getXY[1]
        ctx.lineTo(x,y);
        angle+=120
        getXY = this.rotateLine(x,y,angle,this.length);x=getXY[0],y=getXY[1]
        
        this.coner.right.x = x-5;
        this.coner.right.y = y-5
        ctx.lineTo(x,y);
        angle+=60
        getXY = this.rotateLine(x,y,angle,this.height);x=getXY[0],y=getXY[1]
        this.coner.bottom.x = x-5;
        this.coner.bottom.y = y-5
        ctx.lineTo(x,y);
        angle+=120
        getXY = this.rotateLine(x,y,angle,this.length);x=getXY[0],y=getXY[1]
        ctx.lineTo(x,y);
        ctx.fill()
        ctx.closePath();
    }

    rotateLine(x,y,angle,len){
        let rotate= angle *Math.PI/180
        x+=len*Math.cos(rotate);
        y+=len*Math.sin(rotate);
        return [x,y]
    }
    update(pos = 1){
        if (this.run) {
            if(this.coner.right.x >c.width ){
                this.dirX=(this.dirX==this.dirY)?-1:1,this.dirY=-1
            }
            if(this.coner.left.x <0){
                this.dirX=(this.dirX==this.dirY)?1:-1,this.dirY=1
            }
            let rotate = (this.dirY == 1)?30:150
            let angle =this.dirX*rotate*Math.PI/180
            this.x+=Math.cos(angle)*pos*this.speed
            this.y+=Math.sin(angle)*pos*this.speed    
          
        }    
    }
    stop(){
        this.speed = 0
        if(this.run){ Block.y+=this.length}
        this.run = false
        if(Block.y <= 200){
            this.y+=this.length
        }
       
    }
    changeSize(b){
        Block.score += ((Math.floor(this.x) == Math.floor(b.x) || Math.floor(this.y) ==Math.floor(b.y))?5:1)
        //console.log(Math.floor(this.x) ,Math.floor(b.x) ,Math.floor(this.y) ,Math.floor(b.y));
        if(Math.floor(this.x) == Math.floor(b.x) || Math.floor(this.y) ==Math.floor(b.y)){create_text("PERFECT")}
        if(b.coner.top.x < this.coner.top.x &&b.coner.right.x < this.coner.right.x &&this.dirX!=this.dirY){
            this.height -= abs(Math.floor(this.x)-b.x)
            this.x =b.x;
            this.y = b.y-this.length
            Block.height = this.height
        }
        if(b.coner.top.x > this.coner.top.x &&b.coner.right.x > this.coner.right.x &&this.dirX!=this.dirY){
            this.height -= abs(Math.floor(this.x)-b.x)
            this.draw();
            b.draw();
            let dx = abs(this.x - this.coner.left.x)
            let dx1 = abs(b.x - b.coner.left.x)
            this.x = b.x-(dx1-dx)
            this.draw()
            b.draw()
            let dx2 = abs(b.coner.left.y - this.coner.left.y)
           // console.log(abs(dx2-this.length),b.coner.left.y, this.coner.left.y);
            this.y+=dx2-this.length
            Block.height = this.height
        } 
        if(b.coner.top.x > this.coner.top.x &&b.coner.right.x > this.coner.right.x &&this.dirX==this.dirY){
            this.width -= abs(Math.floor(this.x)-b.x)
            this.x =b.x;
            this.y = b.y-this.length
            Block.width = this.width
        }
        if(b.coner.top.x < this.coner.top.x &&b.coner.right.x < this.coner.right.x &&this.dirX==this.dirY){
            this.width -= abs(Math.floor(this.x)-b.x)
            this.draw();
            b.draw();
            let dx = abs(this.x - this.coner.right.x)
            let dx1 = abs(b.x - b.coner.right.x)
            this.x = b.x+abs(dx1-dx)
            Block.width = this.width
            this.draw()
            b.draw()
            console.log(b.coner.right.x - this.coner.right.x);
            let dx2 = abs(b.coner.right.y - this.coner.right.y)
           // console.log(abs(dx2-this.length),b.coner.left.y, this.coner.left.y);
            this.y+=dx2-this.length
          /*   
           
            */
        } 
        Block.x = this.x;
        Block.y = this.y-this.length
        if(this.height <= 0 || this.width <= 0){
            Block.speed = 0;
            messages.style.display = "block";
            messages.innerText = "Click to Restart";
            create_text("Game Over");
            RunGame = false
        }
       
    }
}
let blocks = []
//let block = new Blocks(1,1,"green");
let hsl = 60
const create_blocks = ()=>{    
    if (Block.height > 0 && Block.speed&&Block.width>0) {
    hsl++
     let dirX = Math.floor(Math.random()*2) == 0?-1:1;
    let dirY = Math.floor(Math.random()*2) == 0?-1:1; 
    /* let dirX =  -1;
    let dirY = -1 */
    blocks.push(new Blocks(dirX,dirY,hsl))
}
}
const blocks_stop = ()=>{
    for (let i = 0; i < blocks.length; i++) {
        blocks[i].stop()
    }
       
     if(blocks[blocks.length-2]&&blocks[blocks.length-1]){
        blocks[blocks.length-1].changeSize(blocks[blocks.length-2])} 
 
} 
create_blocks()
blocks_stop()
create_blocks()
const onClick=()=>{
    messages.style.display = "none"
    Block.speed+=0.0005
    if(!RunGame){
        RunGame=true;
        blocks = [];
        first_block = true;
        Block.speed = 0.2;
        Block.score = 1;
        Block.start = false;
        Block.width = 200;
        Block.height = 200;
        Block.x=c.width/2;
        Block.y=c.height-250;
        hsl = 60
        create_blocks()
    }
    blocks_stop()
    create_blocks()
    if(!Block.start){
        Block.time.start = new Date().getTime();
        Block.start=true;
    }
    
     if(parseInt(Block.highscore) < Block.score){
        Block.show_msg = false
        localStorage.setItem("highscore",Block.score)
        
        if(Block.show_msg){
        create_text("New Score")}
        
     }
     
    Block.highscore=localStorage.getItem("highscore")
    if(Block.speed){
    speed.innerText = `Speed: ${Math.floor((Block.speed/0.2)*10)/10}x`   }  
    highscore.innerText = `HighScore:${Block.highscore}`
}
highscore.innerText = `HighScore:  ${Block.highscore}`
window.addEventListener("keypress",(e)=>{

    if(e.key == " "){
       onClick()
    }
    
})

const handle_blocks = ()=>{
    for (let i = 0; i < blocks.length; i++) {
        const block = blocks[i];
        block.draw()
        block.update()
        if(block.height <=0 || block.width <= 0){
            blocks.splice(i,1)
            i--;
        }
    }
}
const setTimer = ()=>{
    if(Block.time.start){
    Block.time.end = new Date().getTime();
    let time = (Block.time.end - Block.time.start)/1000;
    let sec = Math.floor((Block.time.end - Block.time.start)/1000);
    
    let min = Math.floor(sec/60)
    let txtsec= sec%60,txtmin = min
    if(sec%60 < 10){
        txtsec= "0"+sec%60;
    }
    if(min<10){
        txtmin= "0"+min;
    }
    timer.innerText = `${txtmin}:${txtsec}`}
    
}
const handleScore = ()=>{
    score_num.innerText = Block.score
   
}
let frame_txt = "FPS: 0"
const animate= ()=>{
    ctx.clearRect(0,0,c.width,c.height);
    handle_blocks()
    handle_text()
    if(RunGame){
    setTimer()}
    handleScore()
    if ((new Date().getTime())/1000 - start_time/1000 >= 1.0){
        start_time = new Date().getTime()
        frame_txt = "FPS: "+frame
        frame = 0}
    frame+=1
    FPS.innerText = frame_txt;
    requestAnimationFrame(animate);
}
window.addEventListener("click",onClick)
animate()